### DISCUSSION TITLE
Select Condition

### DISCUSSION THREAD

**Discussion by morellik** [dflib/dflib#283] [Q&A] [answered]
A question about the select condition. I've a record with a mix of fields type:

_private record PDBRecord(String recordName,
             Integer atomNumber,
             String atomName,
             String residueName,
             String chain,
             String residueNumber,
             Double xCoordinate,
             Double yCoordinate,
             Double zCoordinate,
             Float occupancy,
             Float tempFactor,
             String segID,
             String elementType) {
    }_

I fill the DataFrame with values:

_dframe = DataFrame
                .byRow(
                        Extractor.$col(PDBRecord::recordName),
                        Extractor.$col(PDBRecord::atomNumber),
                        Extractor.$col(PDBRecord::atomName),
                        Extractor.$col(PDBRecord::residueName),
                        Extractor.$col(PDBRecord::chain),
                        Extractor.$col(PDBRecord::residueNumber),
                        Extractor.$col(PDBRecord::xCoordinate),
                        Extractor.$col(PDBRecord::yCoordinate),
                        Extractor.$col(PDBRecord::zCoordinate),
                        Extractor.$col(PDBRecord::occupancy),
                        Extractor.$col(PDBRecord::tempFactor),
                        Extractor.$col(PDBRecord::segID),
                        Extractor.$col(PDBRecord::elementType)
                )
                .columnNames("record", "atomNumber", "atomName", "residueName", "chain", "residueNumber",
                        "x", "y", "z", "occupancy", "tempFactor", "segID", "elementType")
                .appender()
                .append(pdbRecords)
                .toDataFrame();_

Now I want to select all rows where the atomNumber is equal to a number. I loop over a list of atomNumber and I would select the  rows where the condition is true:

_for(Map.Entry<String, List<List<String>>> entry: metal2ligands.entrySet()) {
            for (List<String> l : entry.getValue()) {
                 System.out.println(l.get(8));
                df1 = PDBDataFrame.rows(r -> r.get("atomNumber", String.class).equals(l.get(8))).select();
            }
        }_

But, if I use String.class I receive, correctly,  java.lang.ClassCastException: Cannot cast java.lang.Integer to java.lang.String.
But If I use Integer.class the IntelliJ IDE shows that the condition is always false (in fact the df1 becomes empty). So, how can select a field for different value types?

Thanks

**Comment by andrus** [comment]
`r.get("atomNumber", Integer.class)` would be the correct way to reference that field. So you need to check what's stored in the `l` list. When you are debugging the output, try this instead: `System.out.println(l.get(8).getClass().getName())`. And depending on what type you have in that List, you may need to do an explicit conversion either of the List value, or the DataFrame column

**Reply by morellik** [reply]
In the end, the problem was the PDBDataFrame creation.  Another question, how can I append to df1 all the selected rows?
for (List l : entry.getValue()) {
    df1 = PDBDataFrame.rows(r -> r.get("atomNumber", String.class).equals(l.get(8))).select();
}

At the end df1 has only the last selection.

**Reply by morellik** [reply]
I tried 

df1.vConcat(PDBDataFrame.rows(r -> r.get("atomNumber", Integer.class).equals(Integer.parseInt(l.get(8)))).select());

but the result is nothing

df1 is an empty DataFrame defined as DataFrame df1 = DataFrame.empty()

But if I use another DataFrame created before, the concatenation doesn't work. DataFrame df before the cycle:
record atomNumber atomName residueName chain residueNumber      x      y      z occupancy tempFactor segID elementType
------ ---------- -------- ----------- ----- ------------- ------ ------ ------ --------- ---------- ----- -----------
ATOM          722 NE2      HIS         A     94            -7.486 -0.334 13.493       1.0       5.58       N          
ATOM          743 NE2      HIS         A     96             -5.89 -3.503 14.906       1.0       4.07       N          
ATOM          920 ND1      HIS         A     119           -8.295 -2.377 16.601       1.0       4.43       N          
ATOM         1583 SG       CYS         A     206           -2.747  9.466 22.676       1.0       8.73       S          
4 rows x 13 columns

for (List<String> l : entry.getValue()) {
         df.vConcat(PDBDataFrame.rows(r -> r.get("atomNumber", Integer.class).equals(Integer.parseInt(l.get(8)))).select());
}

DataFrame df after the concatenation:

record atomNumber atomName residueName chain residueNumber      x      y      z occupancy tempFactor segID elementType
------ ---------- -------- ----------- ----- ------------- ------ ------ ------ --------- ---------- ----- -----------
ATOM          722 NE2      HIS         A     94            -7.486 -0.334 13.493       1.0       5.58       N          
ATOM          743 NE2      HIS         A     96             -5.89 -3.503 14.906       1.0       4.07       N          
ATOM          920 ND1      HIS         A     119           -8.295 -2.377 16.601       1.0       4.43       N          
ATOM         1583 SG       CYS         A     206           -2.747  9.466 22.676       1.0       8.73       S          
4 rows x 13 columns


If I remove the concatenation, the df1 DataFrame contains the last selection:

record atomNumber atomName residueName chain residueNumber      x      y      z occupancy tempFactor segID elementType
------ ---------- -------- ----------- ----- ------------- ------ ------ ------ --------- ---------- ----- -----------
ATOM          923 NE2      HIS         A     119           -8.993 -3.545 18.296       1.0       4.57       N          
1 row x 13 columns

**Reply by morellik** [reply]
Reading the documentation I understand that I can concatenate two existing DataFrame in a third one. But is there a way to do that in a loop as in the example above?

**Comment by morellik** [comment]
Seems that the following code is working where df is a DataFrame with other data.

 for (List<String> l : entry.getValue()) {
                System.out.println(l.get(8));
                df1 = PDBDataFrame.rows(r -> r.get("atomNumber", Integer.class).equals(Integer.parseInt(l.get(8)))).select();
                df = df.vConcat(df1);
            }

Is it the correct way?

**Reply by andrus** [reply, marked as answer]
Yes, this is fine. If performance is a consideration, you can collect all DataFrames in a `DataFrame[]` first, and then append them all at once. It will be somewhat more efficient:

```java
List<DataFrame> toAppend = ... 
for (List l : entry.getValue()) {
   toAppend.add(...);
}

df = df.vConcat(toAppend.toArray(new DataFrame[0]));
```

**Reply by morellik** [reply]
Wow! I'm starting to love dflib. I hope you'll continue to maintain dflib and offer support.

Thanks you very much.

**Reply by andrus** [reply]
> Wow! I'm starting to love dflib. 

Thanks for the kind words 🙂

> I hope you'll continue to maintain dflib and offer support.

That's the plan. We have a big roadmap, and just announced that we are going to support JJava Jupyter kernel (IMO, an important piece otherwise missing in the Java data processing ecosystem) - https://github.com/SpencerPark/IJava/issues/161
